### PR TITLE
Adding user configurable socket options

### PR DESCRIPTION
**Pull request by nsarras**
Users are now able to set socket options from the datamover config file

```
"socket_options": {
    "connection_delay_interval": 2,
    "connection_time_threshold": 300,
    "message_length": 10,
    "recv_timeout": 60
  },
```
image

Data Mover Put operation ran successfully.

image

### CODE REVIEW

File: dss_datamover/socket_communication.py [hunk 1]
```diff
@@ -42,17 +42,32 @@
...
 class ClientSocket:
 
-    def __init__(self, logger=None, ip_address_family="IPv4"):
+    def __init__(self, config, logger=None, ip_address_family="IPv4"):
         self.logger = logger
+        self.config = config
         self.socket = None
+
+        self.CONNECTION_DELAY_INTERVAL = (
+            self.config.get("socket_options", DEFAULT_CONNECTION_DELAY_INTERVAL).get("connection_delay_interval", DEFAULT_CONNECTION_DELAY_INTERVAL)
```
Comment (grandsuri) [inline dss_datamover/socket_communication.py:59]: This fails if you don't mention **socket_options** in config file

File: dss_datamover/socket_communication.py [hunk 2]
```diff
@@ -77,7 +92,8 @@ def connect(self, host, port):
 
         is_connection_refused = True
         connection_time_start = datetime.now()
-        time_to_sleep = CONNECTION_DELAY_INTERVAL
+        time_to_sleep = self.config.get("socket_options", ).get("connection_delay_interval")
```
Comment (grandsuri) [inline dss_datamover/socket_communication.py:95]: why do you need this code when you are overriding the variable . See below line

File: dss_datamover/socket_communication.py [hunk 3]
```diff
@@ -42,17 +42,32 @@
     "IPv6": socket.AF_INET6
 }
 
-CONNECTION_DELAY_INTERVAL = 2
-CONNECTION_TIME_THRESHOLD = 300  # 5 Minutes, maximum wait time for socket connection.
-MESSAGE_LENGTH = 10  # Message length 10 bytes
-RECV_TIMEOUT = 60  # Wait to receive data from socket for 60 seconds.
+DEFAULT_CONNECTION_DELAY_INTERVAL = 2
+DEFAULT_CONNECTION_TIME_THRESHOLD = 300  # 5 Minutes, maximum wait time for socket connection.
+DEFAULT_MESSAGE_LENGTH = 10  # Message length 10 bytes
```
Comment (grandsuri) [inline dss_datamover/socket_communication.py:47]: change this name to DEFAULT_CLIENT_RESP_HEADER_LENGTH or something meaningful

File: dss_datamover/config/config.json [hunk 4]
```diff
@@ -17,6 +17,12 @@
     "port_status": 6001
   },
   "ip_address_family" : "IPV6",
+  "socket_options": {
+    "connection_delay_interval": 2,
```
Comment (grandsuri) [inline dss_datamover/config/config.json:21]: Better to add some comments for easy understanding

File: dss_datamover/socket_communication.py [hunk 5]
```diff
@@ -216,17 +232,33 @@ def close(self):
...
 
-    def __init__(self, logger=None, ip_address_family="IPv4"):
+    def __init__(self, config, logger=None, ip_address_family="IPv4"):
         self.socket = None
         self.logger = logger
+        self.config = config
         self.client_socket = None
+
+        self.CONNECTION_DELAY_INTERVAL = (
+            self.config.get("socket_options", DEFAULT_CONNECTION_DELAY_INTERVAL).get("connection_delay_interval", DEFAULT_CONNECTION_DELAY_INTERVAL)
```
Comment (grandsuri) [inline dss_datamover/socket_communication.py:242]: Don't use CAPS for variables. Strictly prohibited in python

File: dss_datamover/config/config.json (hunk 4 above)
Comment (nsarras) [reply dss_datamover/config/config.json:21]: I can't add comments to a JSON file

File: dss_datamover/socket_communication.py (hunk 2 above)
Comment (nsarras) [reply dss_datamover/socket_communication.py:95]: removed unnecessary line

File: dss_datamover/config/config.json [hunk 6]
```diff
@@ -17,6 +17,12 @@
...
+    "connection_time_threshold": 300,
+    "message_length": 10,
```
Comment (grandsuri) [inline dss_datamover/config/config.json:23]: this is confusing for the end users. Change it to response header length as you mentioned in the code.
Also lot of these fields are not understandable for end user. Need to have some simple names or documentation for that.

**Review by grandsuri — COMMENTED** [review]
SonarCloud Quality gate failed. Review that code and reduce the nested conditions.

**Review by grandsuri — APPROVED** [review]

**Pull request merged**